File: 2d-bezier-curve.cpp
// 2D Bézier Curve Implementation in OpenGL with Coordinate Axes
#include <GL/glut.h>
#include <stdio.h>
#include <math.h>

// Control points for a cubic Bézier curve
GLfloat ctrlPoints[4][3] = {
    {-4.0, 0.0, 0.0},  // P0
    {-2.0, 4.0, 0.0},  // P1
    { 2.0, 4.0, 0.0},  // P2
    { 4.0, 0.0, 0.0}   // P3
};

// Function to calculate factorial
int factorial(int n) {
    if (n <= 1)
        return 1;
    return n * factorial(n - 1);
}

// Function to calculate combination nCr
int combination(int n, int r) {
    return factorial(n) / (factorial(r) * factorial(n - r));
}

// Calculate a point on Bézier curve using Bernstein basis
void calculateBezierPoint(float t, float* point, int numControlPoints) {
    int i, j;
    float bernstein;
    point[0] = 0.0;
    point[1] = 0.0;
    point[2] = 0.0;

    for (i = 0; i < numControlPoints; i++) {
        bernstein = combination(numControlPoints - 1, i) *
                    pow(t, i) *
                    pow(1 - t, numControlPoints - 1 - i);

        for (j = 0; j < 3; j++) {
            point[j] += ctrlPoints[i][j] * bernstein;
        }
    }
}

// Draw coordinate axes with labels
void drawAxes() {
    // Draw X and Y axes
    glLineWidth(1.5);

    // X-axis (red)
    glColor3f(1.0, 0.0, 0.0);
    glBegin(GL_LINES);
        glVertex3f(-5.0, 0.0, 0.0);
        glVertex3f(5.0, 0.0, 0.0);
    glEnd();

    // X-axis arrow
    glBegin(GL_TRIANGLES);
        glVertex3f(5.0, 0.0, 0.0);
        glVertex3f(4.8, 0.1, 0.0);
        glVertex3f(4.8, -0.1, 0.0);
    glEnd();

    // Y-axis (green)
    glColor3f(0.0, 1.0, 0.0);
    glBegin(GL_LINES);
        glVertex3f(0.0, -5.0, 0.0);
        glVertex3f(0.0, 5.0, 0.0);
    glEnd();

    // Y-axis arrow
    glBegin(GL_TRIANGLES);
        glVertex3f(0.0, 5.0, 0.0);
        glVertex3f(0.1, 4.8, 0.0);
        glVertex3f(-0.1, 4.8, 0.0);
    glEnd();

    // Draw tick marks on axes
    glColor3f(0.5, 0.5, 0.5);

    // X-axis ticks
    for (int i = -4; i <= 4; i++) {
        if (i == 0) continue; // Skip origin
        glBegin(GL_LINES);
            glVertex3f(i, -0.1, 0.0);
            glVertex3f(i, 0.1, 0.0);
        glEnd();
    }

    // Y-axis ticks
    for (int i = -4; i <= 4; i++) {
        if (i == 0) continue; // Skip origin
        glBegin(GL_LINES);
            glVertex3f(-0.1, i, 0.0);
            glVertex3f(0.1, i, 0.0);
        glEnd();
    }

    // Draw labels
    glColor3f(0.0, 0.0, 0.0);
    glRasterPos3f(5.2, 0.0, 0.0);
    glutBitmapCharacter(GLUT_BITMAP_HELVETICA_12, 'X');

    glRasterPos3f(0.0, 5.2, 0.0);
    glutBitmapCharacter(GLUT_BITMAP_HELVETICA_12, 'Y');

    glRasterPos3f(0.2, 0.2, 0.0);
    glutBitmapCharacter(GLUT_BITMAP_HELVETICA_12, 'O');
}

// Draw the Bézier curve
void drawBezierCurve() {
    float t, point[3];

    // First draw the axes
    drawAxes();

    // Draw control points
    glColor3f(0.0, 0.0, 1.0);  // Blue points
    glPointSize(5.0);
    glBegin(GL_POINTS);

    for (int i = 0; i < 4; i++) {
        glVertex3fv(ctrlPoints[i]);
    }

    glEnd();

    // Draw control polygon
    glColor3f(0.5, 0.5, 0.5);  // Gray lines
    glBegin(GL_LINE_STRIP);

    for (int i = 0; i < 4; i++) {
        glVertex3fv(ctrlPoints[i]);
    }

    glEnd();

    // Draw the Bézier curve
    glColor3f(1.0, 0.0, 0.0);  // Red curve
    glLineWidth(2.0);
    glBegin(GL_LINE_STRIP);

    for (t = 0.0; t <= 1.0; t += 0.01) {
        calculateBezierPoint(t, point, 4);
        glVertex3fv(point);
    }

    glEnd();
    glLineWidth(1.0);
}

// Display function
void display() {
    glClear(GL_COLOR_BUFFER_BIT);
    drawBezierCurve();
    glutSwapBuffers();
}

// Reshape function to maintain aspect ratio
void reshape(int w, int h) {
    glViewport(0, 0, w, h);
    glMatrixMode(GL_PROJECTION);
    glLoadIdentity();

    if (w <= h)
        glOrtho(-5.0, 5.0, -5.0 * (GLfloat) h / (GLfloat) w,
                5.0 * (GLfloat) h / (GLfloat) w, -5.0, 5.0);
    else
        glOrtho(-5.0 * (GLfloat) w / (GLfloat) h,
                5.0 * (GLfloat) w / (GLfloat) h, -5.0, 5.0, -5.0, 5.0);

    glMatrixMode(GL_MODELVIEW);
    glLoadIdentity();
}

// Main function
int main(int argc, char** argv) {
    glutInit(&argc, argv);
    glutInitDisplayMode(GLUT_DOUBLE | GLUT_RGB);
    glutInitWindowSize(500, 500);
    glutInitWindowPosition(100, 100);
    glutCreateWindow("2D Bézier Curve with Axes");

    glClearColor(1.0, 1.0, 1.0, 1.0);
    glutDisplayFunc(display);
    glutReshapeFunc(reshape);

    glutMainLoop();
    return 0;
}

File: 2d-bezier-curve.py
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as patches
from math import factorial

# Control points for a cubic Bézier curve (same as in the original code)
ctrl_points = np.array([
    [-4.0, 0.0, 0.0],  # P0
    [-2.0, 4.0, 0.0],  # P1
    [2.0, 4.0, 0.0],   # P2
    [4.0, 0.0, 0.0]    # P3
])

# Function to calculate combination nCr
def combination(n, r):
    return factorial(n) // (factorial(r) * factorial(n - r))

# Calculate a point on Bézier curve using Bernstein basis
def calculate_bezier_point(t, control_points):
    num_control_points = len(control_points)
    point = np.zeros(3)
    
    for i in range(num_control_points):
        bernstein = combination(num_control_points - 1, i) * \
                   (t ** i) * \
                   ((1 - t) ** (num_control_points - 1 - i))
        
        point += control_points[i] * bernstein
    
    return point

# Create figure and axis
fig, ax = plt.subplots(figsize=(10, 10))

# Draw coordinate axes with labels
def draw_axes():
    # X-axis
    ax.arrow(x=-5, y=0, dx=10, dy=0, head_width=0.2, head_length=0.2, 
             fc='r', ec='r', length_includes_head=True)
    
    # Y-axis
    ax.arrow(x=0, y=-5, dx=0, dy=10, head_width=0.2, head_length=0.2, 
             fc='g', ec='g', length_includes_head=True)
    
    # Axis labels
    ax.text(5.2, 0, 'X', color='r', fontsize=12)
    ax.text(0, 5.2, 'Y', color='g', fontsize=12)
    ax.text(0.2, 0.2, 'O', color='black', fontsize=12)
    
    # Draw tick marks
    for i in range(-4, 5):
        if i == 0:
            continue  # Skip origin
        # X-axis ticks
        ax.plot([i, i], [-0.1, 0.1], color='gray')
        # Y-axis ticks
        ax.plot([-0.1, 0.1], [i, i], color='gray')

# Draw the Bézier curve
def draw_bezier_curve():
    # Draw axes
    draw_axes()
    
    # Draw control points
    ax.scatter(ctrl_points[:, 0], ctrl_points[:, 1], color='blue', s=50)
    
    # Draw control polygon
    ax.plot(ctrl_points[:, 0], ctrl_points[:, 1], 'gray', linestyle='-')
    
    # Draw the Bézier curve
    t_values = np.linspace(0, 1, 100)
    curve_points = np.array([calculate_bezier_point(t, ctrl_points) for t in t_values])
    
    ax.plot(curve_points[:, 0], curve_points[:, 1], 'r-', linewidth=2)

# Main function
if __name__ == "__main__":
    # Set equal aspect ratio
    ax.set_aspect('equal')
    
    # Set axis limits
    ax.set_xlim(-5.5, 5.5)
    ax.set_ylim(-5.5, 5.5)
    
    # Set background color
    ax.set_facecolor('white')
    
    # Draw the Bézier curve
    draw_bezier_curve()
    
    # Hide the normal axes
    ax.set_xticks([])
    ax.set_yticks([])
    ax.spines['top'].set_visible(False)
    ax.spines['right'].set_visible(False)
    ax.spines['bottom'].set_visible(False)
    ax.spines['left'].set_visible(False)
    
    plt.title('2D Bézier Curve with Axes')
    plt.tight_layout()
    plt.show()
